tfhe/src/integer/server_key/radix_parallel/scalar_div_mod.rs

```diff
@@ -0,0 +1,480 @@
+        // Keep the same name as in the paper
+        let N = integer_bits;
+
+        assert_ne!(divisor, 0);
+        assert!(precision >= 1 && precision <= N);
+
+        let d = u128::from(divisor);
+        let two_pow_n = 1u128 << N;
+
+        // Find l such that (l-1) < log2(d) <= l
+        let l = d.ilog2() + u32::from(!d.is_power_of_two());
+        let mut shift_post = l;
+
+        // The formula is m_low == 2**(N+l) / d
+        // however N + l may be == to 128, which means doing
+        // 2**(N + l) would not be valid, so the formula has been written in a
+        // way that does not overlflow.
+        //
+        // By writing m_low = (m_low - 2**N) + (2**N)
+        let mut m_low = (two_pow_n * ((1 << l) - d)) / d;
+        m_low += two_pow_n;
+
+        // The formula is (2**(N+l) + 2**(N + l - precision)) / d
+        // again 2**(N+l) could overflow
+        let mut m_high = ((two_pow_n * ((1 << l) - d)) + (1 << (N + l - precision))) / d;
+        m_high += two_pow_n;
+
+        assert!(m_low < m_high);
+
+        loop {
+            let m_low_i = m_low >> 1;
+            let m_high_i = m_high >> 1;
+
+            if m_low_i >= m_high_i || shift_post == 0 {
+                break;
+            }
+
+            m_low = m_low_i;
+            m_high = m_high_i;
+            shift_post -= 1;
+        }
+
+        Self {
+            multiplier: m_high,
+            shift_post,
+        }
+    }
+}
+
+impl ServerKey {
+    /// computes lhs * rhs on the full precision (num_block * 2) and returns
+    /// the most significant num blocks.
+    ///
+    /// In other words, it is like doing
+    /// fn mulhi (a: u32, b: u32) -> u32 {
+    ///    (a as 64 * b as u64) >> 32) as u32
+    /// }
+    fn scalar_mul_high(&self, lhs: &RadixCiphertext, rhs: u128) -> RadixCiphertext {
+        let mut result = lhs.clone();
+        self.extend_radix_with_trivial_zero_blocks_msb_assign(&mut result, lhs.blocks.len());
+        self.scalar_mul_assign_parallelized(&mut result, rhs);
+        self.trim_radix_blocks_lsb_assign(&mut result, lhs.blocks.len());
+        result
+    }
+
+    pub fn unchecked_scalar_div_parallelized<T>(
+        &self,
+        numerator: &RadixCiphertext,
+        divisor: T,
+    ) -> RadixCiphertext
+    where
+        u64: From<T>,
+        T: Numeric,
+    {
+        let numerator_bits = self.key.message_modulus.0.ilog2() * numerator.blocks.len() as u32;
+
+        let divisor = u64::from(divisor);
+        // Rust has a check on all devision, so shall also have one
+        assert_ne!(divisor, 0, "Cannot divide by zero");
+
+        if divisor.is_power_of_two() {
+            // Even in FHE, shifting is faster than multiplying / dividing
+            return self
+                .unchecked_scalar_right_shift_parallelized(numerator, divisor.ilog2() as u64);
+        }
+
+        let log2_divisor = divisor.ilog2() + 1; // We want ceil(log2(divisor))
```

Comment:
that does not work for power of 2 ?

I think we need a function somewhere which does that computation, I'm starting to use it more and more myself

Reply:
Yes but just above we branch out if divisor is a power of two so this is never executed when divisor is a power of two

Reply:
while I agree I would feel better with a function doing the ceil_log2 with those checks in place 🙂 easy to misuse even if we are familiar with it !

Reply:
added ceil_ilog2 to UnsignedIntger